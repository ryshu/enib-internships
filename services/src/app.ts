import express, { Request, Response, NextFunction } from 'express';
import bodyParser from 'body-parser';
import cors from 'cors';
import compression from 'compression'; // compresses requests
import lusca from 'lusca';
import dotenv from 'dotenv';
import flash from 'express-flash';
import path from 'path';
import methodOverride from 'method-override';

// Load environment variables from .env file, where API keys and passwords are configured
dotenv.config({ path: '.env' });

// Loggers
import logger from './utils/logger';
import { APIError } from './utils/error';
import httpStatus from 'http-status-codes';

// router and database
import router from './configs/setup/route';
import './configs/instances/database'; // Only import to setup
import './configs/setup/database'; // Only import to setup

// Create Express server
const app = express();

// Express configuration
app.set('port', process.env.INTERNSHIP_ENIB_API_PORT || 3000);
app.use(compression());
app.use(bodyParser.json({ limit: '1mb' }));
app.use(bodyParser.urlencoded({ extended: true, limit: '1mb' }));
app.use(flash());
app.use(lusca.xframe('SAMEORIGIN'));
app.use(lusca.xssProtection(true));

app.use(express.static(path.join(__dirname, 'public'), { maxAge: 31557600000 }));

const corsOptions = {
    origin: 'http://localhost:4200',
    optionsSuccessStatus: 200,
};
app.use(cors(corsOptions));

app.use(router);

/**
 * Error handler setup
 */
app.use(methodOverride());
app.use((err: APIError | Error, _req: Request, res: Response, _next: NextFunction) => {
    if (err instanceof APIError) {
        return res.status(err.status).json(err);
    } else {
        logger.error(err);
        return res.status(httpStatus.INTERNAL_SERVER_ERROR).send(`Internal server error`);
    }
});

export default app;
