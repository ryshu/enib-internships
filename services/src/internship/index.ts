export * from './internship';
export * from './internship.mode';
export * from './internship.result';
