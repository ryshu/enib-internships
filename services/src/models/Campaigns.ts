import * as Sequelize from 'sequelize';

import database from '../configs/instances/database';

import MPS from './MentoringPropositions';
import Internships from './Internships';

class Campaigns extends Sequelize.Model implements ICampaignEntity {
    public static associations: {
        propositions: Sequelize.Association<Campaigns, MPS>;
        availbleInternships: Sequelize.Association<Campaigns, Internships>;
        validatedInternships: Sequelize.Association<Campaigns, Internships>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public name: string;
    public description: string;
    public startAt: number;
    public endAt: number;
    public semester: string;
    public maxProposition: number;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    public getPropositions: Sequelize.HasManyGetAssociationsMixin<MPS>;
    public addProposition: Sequelize.HasManyAddAssociationMixin<MPS, MPS['id']>;
    public createProposition: Sequelize.HasManyCreateAssociationMixin<IMentoringPropositionEntity>;
    public hasProposition: Sequelize.HasManyHasAssociationMixin<MPS, MPS['id']>;
    public countPropositions: Sequelize.HasManyCountAssociationsMixin;

    // AvailableInternships
    public getAvailableInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addAvailableInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createAvailableInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasAvailableInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countAvailableInternships: Sequelize.HasManyCountAssociationsMixin;

    // ValidatedInternships
    public getValidatedInternships: Sequelize.HasManyGetAssociationsMixin<Internships>;
    public addValidatedInternship: Sequelize.HasManyAddAssociationMixin<Internships, Internships['id']>;
    public createValidatedInternship: Sequelize.HasManyCreateAssociationMixin<IInternshipEntity>;
    public hasValidatedInternship: Sequelize.HasManyHasAssociationMixin<Internships, Internships['id']>;
    public countValidatedInternships: Sequelize.HasManyCountAssociationsMixin;

    public readonly propositions?: MPS[];
    public readonly availableInternships?: Internships[];
    public readonly validatedInternships?: Internships[];
}

Campaigns.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        name: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        description: {
            type: new Sequelize.DataTypes.STRING(),
            allowNull: false,
        },
        startAt: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            allowNull: true,
            defaultValue: null,
        },
        endAt: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            allowNull: true,
            defaultValue: null,
        },
        semester: {
            type: Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        maxProposition: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            defaultValue: 0,
            allowNull: true,
        },
    },
    {
        tableName: 'campaigns',
        sequelize: database,
    },
);

export default Campaigns;
