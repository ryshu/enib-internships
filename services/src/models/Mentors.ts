import * as Sequelize from 'sequelize';

import database from '../configs/instances/database';
import Campaigns from './Campaigns';

class Mentors extends Sequelize.Model implements IMentorEntity {
    public static associations: {
        campaigns: Sequelize.Association<Mentors, Campaigns>;
    };

    public id!: number; // Note that the `null assertion` `!` is required in strict mode.

    public firstName: string;
    public lastName: string;
    public email: string;
    public role: 'default' | 'admin';

    // Campaigns
    public getCampaigns: Sequelize.BelongsToManyGetAssociationsMixin<Campaigns>;
    public addCampaign: Sequelize.BelongsToManyAddAssociationMixin<Campaigns, Campaigns['id']>;
    public hasCampaign: Sequelize.BelongsToManyHasAssociationMixin<Campaigns, Campaigns['id']>;
    public countCampaigns: Sequelize.BelongsToManyCountAssociationsMixin;

    // timestamps!
    public readonly createdAt!: Date;
    public readonly updatedAt!: Date;

    public readonly campaigns?: Campaigns[];
}

Mentors.init(
    {
        id: {
            type: Sequelize.DataTypes.INTEGER.UNSIGNED,
            autoIncrement: true,
            primaryKey: true,
        },
        firstName: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        lastName: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        email: {
            type: new Sequelize.DataTypes.STRING(128),
            allowNull: false,
        },
        role: {
            type: new Sequelize.DataTypes.STRING(40),
            allowNull: true,
            defaultValue: 'default',
        },
    },
    {
        tableName: 'mentors',
        sequelize: database,
    },
);

export default Mentors;
