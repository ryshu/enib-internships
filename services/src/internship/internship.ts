import moment from 'moment';

import { INTERNSHIP_RESULT, isInternshipResult } from './internship.result';
import { INTERNSHIP_MODE, isInternshipMode } from './internship.mode';

import Internships from '../models/sequelize/Internships';
import Students from '../models/sequelize/Students';
import Businesses from '../models/sequelize/Businesses';
import InternshipTypes from '../models/sequelize/InternshipTypes';
import Campaigns from '../models/sequelize/Campaigns';
import Mentors from '../models/sequelize/Mentors';
import MentoringPropositions from '../models/sequelize/MentoringPropositions';
import Files from '../models/sequelize/Files';

import { APIError } from '../utils/error';
import { IInternshipEntity, ICampaignEntity } from '../declarations';

import cache from '../statistics/singleton';

function isInt(t: number): t is number {
    return t && !Number.isNaN(Number(t));
}

export class InternshipHandler {
    private _internship: Internships;

    public constructor(internship: Internships) {
        this._internship = internship;
    }

    public toWaiting(): Promise<InternshipHandler> {
        return new Promise(async (resolve, reject) => {
            try {
                if (this.state !== INTERNSHIP_MODE.PUBLISHED) {
                    return reject(this._buildForbiddenError(INTERNSHIP_MODE.WAITING));
                }

                this._internship.publishAt = 0;
                this._internship.state = INTERNSHIP_MODE.WAITING;
                await this.save();

                this._updateStatistics(INTERNSHIP_MODE.PUBLISHED, INTERNSHIP_MODE.WAITING);

                resolve(this);
            } catch (error) {
                reject(error);
            }
        });
    }

    public toPublished(): Promise<InternshipHandler> {
        return new Promise(async (resolve, reject) => {
            try {
                if (this.state !== INTERNSHIP_MODE.WAITING) {
                    return reject(this._buildForbiddenError(INTERNSHIP_MODE.PUBLISHED));
                }

                this._internship.publishAt = moment().valueOf();
                this._internship.state = INTERNSHIP_MODE.PUBLISHED;

                this._updateStatistics(INTERNSHIP_MODE.WAITING, INTERNSHIP_MODE.PUBLISHED);

                await this.save();

                resolve(this);
            } catch (error) {
                reject(error);
            }
        });
    }

    public toAttributedStudent(studentId: number): Promise<InternshipHandler> {
        return new Promise(async (resolve, reject) => {
            try {
                if (this.state !== INTERNSHIP_MODE.PUBLISHED) {
                    return reject(this._buildForbiddenError(INTERNSHIP_MODE.ATTRIBUTED_STUDENT));
                }

                if (!isInt(studentId)) {
                    // TODO: Reject params val
                    return resolve(this);
                }
                const student = await Students.findByPk(studentId);
                if (!student) {
                    return resolve(this);
                }

                await this._internship.setStudent(student);
                this._internship.state = INTERNSHIP_MODE.ATTRIBUTED_STUDENT;
                await this.save();

                this._updateStatistics(
                    INTERNSHIP_MODE.PUBLISHED,
                    INTERNSHIP_MODE.ATTRIBUTED_STUDENT,
                );

                resolve(this);
            } catch (error) {
                reject(error);
            }
        });
    }

    public toCampaignAvailable(campaignId: number): Promise<InternshipHandler> {
        return new Promise(async (resolve, reject) => {
            try {
                if (this.state !== INTERNSHIP_MODE.ATTRIBUTED_STUDENT) {
                    return reject(this._buildForbiddenError(INTERNSHIP_MODE.AVAILABLE_CAMPAIGN));
                }

                if (!isInt(campaignId)) {
                    // TODO: Reject params val
                    return resolve(this);
                }

                const campaign = await Campaigns.findByPk(campaignId);
                if (!campaign) {
                    return resolve(this);
                }

                if (this._internship.validatedCampaign) {
                    await campaign.removeValidatedInternships(this._internship);
                }
                await campaign.addAvailableInternship(this._internship);
                this._internship.state = INTERNSHIP_MODE.AVAILABLE_CAMPAIGN;
                await this.save();

                this._updateStatistics(
                    INTERNSHIP_MODE.ATTRIBUTED_STUDENT,
                    INTERNSHIP_MODE.AVAILABLE_CAMPAIGN,
                );

                resolve();
            } catch (error) {
                reject(error);
            }
        });
    }

    public toAttributedMentor(mentorId: number): Promise<InternshipHandler> {
        return new Promise(async (resolve, reject) => {
            try {
                if (this.state !== INTERNSHIP_MODE.AVAILABLE_CAMPAIGN) {
                    return reject(this._buildForbiddenError(INTERNSHIP_MODE.ATTRIBUTED_MENTOR));
                }

                if (
                    !isInt(mentorId) ||
                    !this._internship.availableCampaign ||
                    !isInt((this._internship.availableCampaign as ICampaignEntity).id)
                ) {
                    // TODO: Reject params val
                    return resolve(this);
                }

                const campaign = await Campaigns.findByPk(
                    (this._internship.availableCampaign as ICampaignEntity).id,
                );
                if (!campaign) {
                    return resolve(this);
                }

                const mentor = await Mentors.findByPk(mentorId);
                if (!mentor) {
                    return resolve(this);
                }

                await this._internship.setMentor(mentor);

                await campaign.removeAvailableInternships(this._internship);
                await campaign.addValidatedInternship(this._internship);

                this._internship.state = INTERNSHIP_MODE.ATTRIBUTED_MENTOR;
                await this.save();

                this._updateStatistics(
                    INTERNSHIP_MODE.AVAILABLE_CAMPAIGN,
                    INTERNSHIP_MODE.ATTRIBUTED_MENTOR,
                );

                resolve(this);
            } catch (error) {
                reject(error);
            }
        });
    }

    public toRunning(endAt?: number): Promise<InternshipHandler> {
        return new Promise(async (resolve, reject) => {
            try {
                if (this.state !== INTERNSHIP_MODE.ATTRIBUTED_MENTOR) {
                    throw this._buildForbiddenError(INTERNSHIP_MODE.RUNNING);
                }
                this._internship.state = INTERNSHIP_MODE.RUNNING;
                this._internship.startAt = moment().valueOf();

                // Setup end date if existing and is number
                if (isInt(endAt)) {
                    this._internship.endAt = moment(endAt).valueOf();
                }

                await this.save();

                this._updateStatistics(INTERNSHIP_MODE.ATTRIBUTED_MENTOR, INTERNSHIP_MODE.RUNNING);

                resolve(this);
            } catch (error) {
                reject(error);
            }
        });
    }

    public toValidation(): Promise<InternshipHandler> {
        return new Promise(async (resolve, reject) => {
            try {
                if (this.state !== INTERNSHIP_MODE.RUNNING) {
                    throw this._buildForbiddenError(INTERNSHIP_MODE.VALIDATION);
                }

                this._internship.state = INTERNSHIP_MODE.VALIDATION;

                await this.save();

                this._updateStatistics(INTERNSHIP_MODE.RUNNING, INTERNSHIP_MODE.VALIDATION);
                resolve(this);
            } catch (error) {
                reject(error);
            }
        });
    }

    public archive(result?: INTERNSHIP_RESULT): Promise<InternshipHandler> {
        return new Promise(async (resolve, reject) => {
            try {
                const prev = this._internship.state;
                this._internship.state = INTERNSHIP_MODE.ARCHIVED;
                this._internship.result = isInternshipResult(result)
                    ? INTERNSHIP_RESULT.VALIDATED
                    : INTERNSHIP_RESULT.UNKNOWN;
                await this.save();

                this._updateStatistics(prev, INTERNSHIP_MODE.ARCHIVED);

                resolve(this);
            } catch (error) {
                reject(error);
            }
        });
    }

    public save(): Promise<InternshipHandler> {
        return new Promise(async (resolve, reject) => {
            try {
                await this._internship.save();
                await this._sync();
                resolve(this);
            } catch (error) {
                reject(error);
            }
        });
    }

    public toJSON(): IInternshipEntity {
        return this._internship.toJSON() as IInternshipEntity;
    }

    public get state() {
        return this._internship.state;
    }

    public get nextState() {
        return this._nextState(this.state);
    }

    private async _sync() {
        this._internship = (await Internships.findByPk(this._internship.id, {
            include: [
                { model: Businesses, as: 'business' },
                { model: InternshipTypes, as: 'category' },
                { model: Campaigns, as: 'availableCampaign' },
                { model: Campaigns, as: 'validatedCampaign' },
                { model: Mentors, as: 'mentor' },
                { model: MentoringPropositions, as: 'propositions' },
                { model: Students, as: 'student' },
                { model: Files, as: 'files' },
            ],
        })) as Internships;
    }

    private _nextState(tested: INTERNSHIP_MODE) {
        let mode: INTERNSHIP_MODE;
        switch (tested) {
            case INTERNSHIP_MODE.ARCHIVED:
                mode = INTERNSHIP_MODE.ARCHIVED;
                break;
            case INTERNSHIP_MODE.ATTRIBUTED_MENTOR:
                mode = INTERNSHIP_MODE.RUNNING;
                break;
            case INTERNSHIP_MODE.ATTRIBUTED_STUDENT:
                mode = INTERNSHIP_MODE.AVAILABLE_CAMPAIGN;
                break;
            case INTERNSHIP_MODE.AVAILABLE_CAMPAIGN:
                mode = INTERNSHIP_MODE.ATTRIBUTED_MENTOR;
                break;
            case INTERNSHIP_MODE.PUBLISHED:
                mode = INTERNSHIP_MODE.ATTRIBUTED_STUDENT;
                break;
            case INTERNSHIP_MODE.RUNNING:
                mode = INTERNSHIP_MODE.VALIDATION;
                break;
            case INTERNSHIP_MODE.VALIDATION:
                mode = INTERNSHIP_MODE.ARCHIVED;
                break;
            case INTERNSHIP_MODE.WAITING:
                mode = INTERNSHIP_MODE.PUBLISHED;
                break;
            default:
                break;
        }

        return mode;
    }

    private _buildForbiddenError(next: INTERNSHIP_MODE) {
        return new APIError(
            `Couldn't change internship from '${this.state}' to '${
                isInternshipMode(next) ? next : 'unknown_mode'
            }', expected next mode is '${this.nextState}'`,
            400,
            12000,
        );
    }

    private _updateStatistics(prev: INTERNSHIP_MODE, next: INTERNSHIP_MODE) {
        let campaignId: number;

        if (this._internship.availableCampaign) {
            campaignId =
                typeof this._internship.availableCampaign === 'string'
                    ? this._internship.availableCampaign
                    : (this._internship.availableCampaign as Campaigns).id;
        } else if (this._internship.validatedCampaign) {
            campaignId =
                typeof this._internship.validatedCampaign === 'string'
                    ? this._internship.validatedCampaign
                    : (this._internship.validatedCampaign as Campaigns).id;
        }
        cache.stateChange(next, prev, campaignId);
    }
}
