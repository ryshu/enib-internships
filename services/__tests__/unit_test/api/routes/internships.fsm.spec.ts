import request from 'supertest';

// Include app for super agent, not server
import app from '../../../../src/app';

// Include database connection
import dbSetup from '../../../../src/configs/setup/database';

// Import model for pre-operation before asserting API methods
import Internships from '../../../../src/models/sequelize/Internships';
import Students from '../../../../src/models/sequelize/Students';
import Campaigns from '../../../../src/models/sequelize/Campaigns';
import Mentors from '../../../../src/models/sequelize/Mentors';

import { INTERNSHIP_MODE, INTERNSHIP_RESULT } from '../../../../src/internship';

import {
    defaultCampaigns,
    defaultInternships,
    defaultStudents,
    defaultMentors,
} from '../../../../__mocks__/mockData';

const baseURL = `/api/${process.env.INTERNSHIP_ENIB_API_VERSION}`;

jest.setTimeout(30000);

beforeAll((done) => {
    dbSetup.then(() => done()).catch((e) => done(e));
});

describe('POST /internships/:id/fsm', () => {
    beforeEach(async () => {
        // Remove all
        await Internships.destroy({ where: {} });
    });

    it('NoInternships_204', async () => {
        const QUERY = { state: INTERNSHIP_MODE.PUBLISHED };
        const RESPONSE = await request(app)
            .post(`${baseURL}/internships/10/fsm`)
            .send(QUERY);
        expect(RESPONSE.status).toBe(204);
    });

    it('BadRequest_400', async () => {
        const RESPONSE = await request(app).post(`${baseURL}/internships/{falseEncoding}/fsm`);
        expect(RESPONSE.status).toBe(400);
    });

    it('WrongBody_MissingResult_400', async () => {
        const QUERY = { state: INTERNSHIP_MODE.ARCHIVED };
        const RESPONSE = await request(app)
            .post(`${baseURL}/internships/10/fsm`)
            .send(QUERY);
        expect(RESPONSE.status).toBe(400);
    });

    it('WrongBody_MissingCampaignId_400', async () => {
        const QUERY = { state: INTERNSHIP_MODE.AVAILABLE_CAMPAIGN };
        const RESPONSE = await request(app)
            .post(`${baseURL}/internships/10/fsm`)
            .send(QUERY);
        expect(RESPONSE.status).toBe(400);
    });

    it('WrongBody_MissingMentorId_400', async () => {
        const QUERY = { state: INTERNSHIP_MODE.ATTRIBUTED_MENTOR };
        const RESPONSE = await request(app)
            .post(`${baseURL}/internships/10/fsm`)
            .send(QUERY);
        expect(RESPONSE.status).toBe(400);
    });

    it('WrongBody_MissingStudentId_400', async () => {
        const QUERY = { state: INTERNSHIP_MODE.ATTRIBUTED_STUDENT };
        const RESPONSE = await request(app)
            .post(`${baseURL}/internships/10/fsm`)
            .send(QUERY);
        expect(RESPONSE.status).toBe(400);
    });

    it('ChangeState_ARCHIVED_200', async () => {
        const QUERY = { state: INTERNSHIP_MODE.ARCHIVED, result: INTERNSHIP_RESULT.CANCELED };
        const VALID_INTERNSHIP = defaultInternships();

        const CREATED = await Internships.create(VALID_INTERNSHIP);
        const RESPONSE = await request(app)
            .post(`${baseURL}/internships/${CREATED.id}/fsm`)
            .send(QUERY);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });

    it('ChangeState_PUBLISHED_200', async () => {
        const QUERY = { state: INTERNSHIP_MODE.PUBLISHED };
        const VALID_INTERNSHIP = defaultInternships();
        VALID_INTERNSHIP.state = INTERNSHIP_MODE.WAITING;

        const CREATED = await Internships.create(VALID_INTERNSHIP);
        const RESPONSE = await request(app)
            .post(`${baseURL}/internships/${CREATED.id}/fsm`)
            .send(QUERY);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            publishAt: expect.any(Number),
            id: expect.any(Number),
        });
    });

    it('ChangeState_ATTRIBUTED_STUDENT_200', async () => {
        const STUDENT = await Students.create(defaultStudents());

        const QUERY = { state: INTERNSHIP_MODE.ATTRIBUTED_STUDENT, studentId: STUDENT.id };
        const VALID_INTERNSHIP = defaultInternships();
        VALID_INTERNSHIP.state = INTERNSHIP_MODE.PUBLISHED;

        const CREATED = await Internships.create(VALID_INTERNSHIP);
        const RESPONSE = await request(app)
            .post(`${baseURL}/internships/${CREATED.id}/fsm`)
            .send(QUERY);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            student: {
                createdAt: expect.any(String),
                updatedAt: expect.any(String),
            },
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });

    it('ChangeState_CAMPAIGN_AVAILABLE_200', async () => {
        const CAMPAIGN = await Campaigns.create(defaultCampaigns());

        const QUERY = { state: INTERNSHIP_MODE.AVAILABLE_CAMPAIGN, campaignId: CAMPAIGN.id };
        const VALID_INTERNSHIP = defaultInternships();
        VALID_INTERNSHIP.state = INTERNSHIP_MODE.ATTRIBUTED_STUDENT;

        const CREATED = await Internships.create(VALID_INTERNSHIP);
        const RESPONSE = await request(app)
            .post(`${baseURL}/internships/${CREATED.id}/fsm`)
            .send(QUERY);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            availableCampaign: {
                createdAt: expect.any(String),
                updatedAt: expect.any(String),
            },
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });

    it('ChangeState_ATTRIBUTED_MENTOR_200', async () => {
        const MENTOR = await Mentors.create(defaultMentors());

        const QUERY = { state: INTERNSHIP_MODE.ATTRIBUTED_MENTOR, mentorId: MENTOR.id };
        const VALID_INTERNSHIP = defaultInternships();
        VALID_INTERNSHIP.state = INTERNSHIP_MODE.AVAILABLE_CAMPAIGN;
        VALID_INTERNSHIP.availableCampaign = defaultCampaigns();

        const CREATED = await Internships.create(VALID_INTERNSHIP, {
            include: [{ association: Internships.associations.availableCampaign }],
        });
        const RESPONSE = await request(app)
            .post(`${baseURL}/internships/${CREATED.id}/fsm`)
            .send(QUERY);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            validatedCampaign: {
                createdAt: expect.any(String),
                updatedAt: expect.any(String),
            },
            mentor: {
                createdAt: expect.any(String),
                updatedAt: expect.any(String),
            },
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });

    it('ChangeState_RUNNING_200', async () => {
        const QUERY = { state: INTERNSHIP_MODE.RUNNING, endAt: 21356456125645 };
        const VALID_INTERNSHIP = defaultInternships();
        VALID_INTERNSHIP.state = INTERNSHIP_MODE.ATTRIBUTED_MENTOR;

        const CREATED = await Internships.create(VALID_INTERNSHIP);
        const RESPONSE = await request(app)
            .post(`${baseURL}/internships/${CREATED.id}/fsm`)
            .send(QUERY);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            startAt: expect.any(Number),
            id: expect.any(Number),
        });
    });

    it('ChangeState_VALIDATION_200', async () => {
        const QUERY = { state: INTERNSHIP_MODE.VALIDATION };
        const VALID_INTERNSHIP = defaultInternships();
        VALID_INTERNSHIP.state = INTERNSHIP_MODE.RUNNING;

        const CREATED = await Internships.create(VALID_INTERNSHIP);
        const RESPONSE = await request(app)
            .post(`${baseURL}/internships/${CREATED.id}/fsm`)
            .send(QUERY);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });

    it('ChangeState_WAITING_200', async () => {
        const QUERY = { state: INTERNSHIP_MODE.WAITING };
        const VALID_INTERNSHIP = defaultInternships();
        VALID_INTERNSHIP.state = INTERNSHIP_MODE.PUBLISHED;

        const CREATED = await Internships.create(VALID_INTERNSHIP);
        const RESPONSE = await request(app)
            .post(`${baseURL}/internships/${CREATED.id}/fsm`)
            .send(QUERY);
        expect(RESPONSE.status).toBe(200);
        expect(RESPONSE.body).toMatchSnapshot({
            createdAt: expect.any(String),
            updatedAt: expect.any(String),
            id: expect.any(Number),
        });
    });
});
