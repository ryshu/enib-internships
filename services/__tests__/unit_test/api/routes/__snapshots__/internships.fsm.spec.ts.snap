// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /internships/:id/fsm ChangeState_ARCHIVED_200 1`] = `
Object {
  "additional": null,
  "address": null,
  "availableCampaign": null,
  "business": null,
  "category": null,
  "city": "Brest",
  "country": "France",
  "createdAt": Any<String>,
  "description": "Stage",
  "endAt": null,
  "files": Array [],
  "id": Any<Number>,
  "isInternshipAbroad": true,
  "mentor": null,
  "postalCode": null,
  "propositions": Array [],
  "publishAt": null,
  "result": "validated",
  "startAt": null,
  "state": "archived",
  "student": null,
  "subject": "test",
  "updatedAt": Any<String>,
  "validatedCampaign": null,
}
`;

exports[`POST /internships/:id/fsm ChangeState_ATTRIBUTED_MENTOR_200 1`] = `
Object {
  "additional": null,
  "address": null,
  "availableCampaign": null,
  "business": null,
  "category": null,
  "city": "Brest",
  "country": "France",
  "createdAt": Any<String>,
  "description": "Stage",
  "endAt": null,
  "files": Array [],
  "id": Any<Number>,
  "isInternshipAbroad": true,
  "mentor": Object {
    "createdAt": Any<String>,
    "email": "john.doe@enib.fr",
    "firstName": "John",
    "id": 2,
    "lastName": "Doe",
    "role": "default",
    "updatedAt": Any<String>,
  },
  "postalCode": null,
  "propositions": Array [],
  "publishAt": null,
  "result": "unknown",
  "startAt": null,
  "state": "attributed_mentor",
  "student": null,
  "subject": "test",
  "updatedAt": Any<String>,
  "validatedCampaign": Object {
    "createdAt": Any<String>,
    "description": "A description to test",
    "endAt": 20191012,
    "id": 2,
    "isPublish": true,
    "maxProposition": 2,
    "name": "test",
    "semester": "S5",
    "startAt": 0,
    "updatedAt": Any<String>,
  },
}
`;

exports[`POST /internships/:id/fsm ChangeState_ATTRIBUTED_STUDENT_200 1`] = `
Object {
  "additional": null,
  "address": null,
  "availableCampaign": null,
  "business": null,
  "category": null,
  "city": "Brest",
  "country": "France",
  "createdAt": Any<String>,
  "description": "Stage",
  "endAt": null,
  "files": Array [],
  "id": Any<Number>,
  "isInternshipAbroad": true,
  "mentor": null,
  "postalCode": null,
  "propositions": Array [],
  "publishAt": null,
  "result": "unknown",
  "startAt": null,
  "state": "attributed_student",
  "student": Object {
    "createdAt": Any<String>,
    "email": "john.doe@enib.fr",
    "firstName": "John",
    "id": 1,
    "lastName": "Doe",
    "semester": "S10",
    "updatedAt": Any<String>,
  },
  "subject": "test",
  "updatedAt": Any<String>,
  "validatedCampaign": null,
}
`;

exports[`POST /internships/:id/fsm ChangeState_CAMPAIGN_AVAILABLE_200 1`] = `
Object {
  "additional": null,
  "address": null,
  "availableCampaign": Object {
    "createdAt": Any<String>,
    "description": "A description to test",
    "endAt": 20191012,
    "id": 1,
    "isPublish": true,
    "maxProposition": 2,
    "name": "test",
    "semester": "S5",
    "startAt": 0,
    "updatedAt": Any<String>,
  },
  "business": null,
  "category": null,
  "city": "Brest",
  "country": "France",
  "createdAt": Any<String>,
  "description": "Stage",
  "endAt": null,
  "files": Array [],
  "id": Any<Number>,
  "isInternshipAbroad": true,
  "mentor": null,
  "postalCode": null,
  "propositions": Array [],
  "publishAt": null,
  "result": "unknown",
  "startAt": null,
  "state": "available_campaign",
  "student": null,
  "subject": "test",
  "updatedAt": Any<String>,
  "validatedCampaign": null,
}
`;

exports[`POST /internships/:id/fsm ChangeState_PUBLISHED_200 1`] = `
Object {
  "additional": null,
  "address": null,
  "availableCampaign": null,
  "business": null,
  "category": null,
  "city": "Brest",
  "country": "France",
  "createdAt": Any<String>,
  "description": "Stage",
  "endAt": null,
  "files": Array [],
  "id": Any<Number>,
  "isInternshipAbroad": true,
  "mentor": null,
  "postalCode": null,
  "propositions": Array [],
  "publishAt": Any<Number>,
  "result": "unknown",
  "startAt": null,
  "state": "published",
  "student": null,
  "subject": "test",
  "updatedAt": Any<String>,
  "validatedCampaign": null,
}
`;

exports[`POST /internships/:id/fsm ChangeState_RUNNING_200 1`] = `
Object {
  "additional": null,
  "address": null,
  "availableCampaign": null,
  "business": null,
  "category": null,
  "city": "Brest",
  "country": "France",
  "createdAt": Any<String>,
  "description": "Stage",
  "endAt": 21356456125645,
  "files": Array [],
  "id": Any<Number>,
  "isInternshipAbroad": true,
  "mentor": null,
  "postalCode": null,
  "propositions": Array [],
  "publishAt": null,
  "result": "unknown",
  "startAt": Any<Number>,
  "state": "running",
  "student": null,
  "subject": "test",
  "updatedAt": Any<String>,
  "validatedCampaign": null,
}
`;

exports[`POST /internships/:id/fsm ChangeState_VALIDATION_200 1`] = `
Object {
  "additional": null,
  "address": null,
  "availableCampaign": null,
  "business": null,
  "category": null,
  "city": "Brest",
  "country": "France",
  "createdAt": Any<String>,
  "description": "Stage",
  "endAt": null,
  "files": Array [],
  "id": Any<Number>,
  "isInternshipAbroad": true,
  "mentor": null,
  "postalCode": null,
  "propositions": Array [],
  "publishAt": null,
  "result": "unknown",
  "startAt": null,
  "state": "archived",
  "student": null,
  "subject": "test",
  "updatedAt": Any<String>,
  "validatedCampaign": null,
}
`;

exports[`POST /internships/:id/fsm ChangeState_WAITING_200 1`] = `
Object {
  "additional": null,
  "address": null,
  "availableCampaign": null,
  "business": null,
  "category": null,
  "city": "Brest",
  "country": "France",
  "createdAt": Any<String>,
  "description": "Stage",
  "endAt": null,
  "files": Array [],
  "id": Any<Number>,
  "isInternshipAbroad": true,
  "mentor": null,
  "postalCode": null,
  "propositions": Array [],
  "publishAt": 0,
  "result": "unknown",
  "startAt": null,
  "state": "waiting",
  "student": null,
  "subject": "test",
  "updatedAt": Any<String>,
  "validatedCampaign": null,
}
`;
